gh-94821: Fix autobind of empty unix domain address

nirs opened the pull request:
When binding a unix socket to an empty address on Linux, the socket is
automatically bound to an available address in the abstract namespace.

    >>> s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    >>> s.bind("")
    >>> s.getsockname()
    b'\x0075499'

Since python 3.9, the socket is bound to the one address:

    >>> s.getsockname()
    b'\x00'

And trying to bind multiple sockets will fail with:

    Traceback (most recent call last):
      File "/home/nsoffer/src/cpython/Lib/test/test_socket.py", line 5553, in testAutobind
        s2.bind("")
    OSError: [Errno 98] Address already in use

Added 2 tests:
- Auto binding empty address on Linux
- Failing to bind an empty address on other platforms

Fixes f6b3a07b7df6 (bpo-44493: Add missing terminated NUL in sockaddr_un's length (GH-26866)

<!-- gh-issue-number: gh-94821 -->
* Issue: gh-94821
<!-- /gh-issue-number -->

nirs committed:
gh-94821: Fix autobind of empty unix domain address

When binding a unix socket to an empty address on Linux, the socket is
automatically bound to an available address in the abstract namespace.

    >>> s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    >>> s.bind("")
    >>> s.getsockname()
    b'\x0075499'

Since python 3.9, the socket is bound to the one address:

    >>> s.getsockname()
    b'\x00'

And trying to bind multiple sockets will fail with:

    Traceback (most recent call last):
      File "/home/nsoffer/src/cpython/Lib/test/test_socket.py", line 5553, in testAutobind
        s2.bind("")
    OSError: [Errno 98] Address already in use

Added 2 tests:
- Auto binding empty address on Linux
- Failing to bind an empty address on other platforms

Fixes f6b3a07b7df6 (bpo-44493: Add missing terminated NUL in sockaddr_un's length (GH-26866)

serhiy-storchaka commented on Modules/socketmodule.c:
```diff
@@ -1724,8 +1724,10 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
 #ifdef __linux__
-        if (path.len > 0 && *(const char *)path.buf == 0) {
-            /* Linux abstract namespace extension */
+        if (path.len == 0 || (path.len > 0 && *(const char *)path.buf == 0)) {
```
It can be simpler (len is always >= 0):
```suggestion
        if (path.len == 0 || *(const char *)path.buf == 0) {
```

serhiy-storchaka commented on Lib/test/test_socket.py:
```diff
@@ -5543,6 +5543,17 @@ def testBytearrayName(self):
+            self.assertTrue(s1.getsockname().startswith(b"\x00"))
+            with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as s2:
+                s2.bind("")
+                self.assertTrue(s2.getsockname().startswith(b"\x00"))
```
Please add a comment explaining why the test looks repeated?

serhiy-storchaka commented on Lib/test/test_socket.py:
```diff
@@ -5543,6 +5543,17 @@ def testBytearrayName(self):
+        # in the abstract namespace.
+        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as s1:
+            s1.bind("")
+            self.assertTrue(s1.getsockname().startswith(b"\x00"))
```
Maybe add also a test that the name is longer than 1 character: `self.assertGreater(len(s1.getsockname()), 1)`? Or even longer than 5 characters?

You can also test that names of two sockets are different.

nirs replied:
The names must be different otherwise bind would fail with "address in use" but adding a test will make this more clear.

nirs committed:
Address Serhiy comments

- Simplify condition since path.len always >= 0
- Make it more clear why we bind to "" twice
- Match abstract address pattern
- Test that sockets are bound to different addresses

serhiy-storchaka reviewed (approved):
:+1:
